### PR TITLE
Arsenal - Fix `FUNC(removeVirtualItems)` for JIP players

### PR DESCRIPTION
**Pull request by johnb432**
**When merged this pull request will:**
- Title.
- Fix for https://discord.com/channels/976165959041679380/1174212946793074768.
- The arsenal now refreshes globally if `FUNC(addVirtualItems)` or `FUNC(removeVirtualItems)` was applied globally.

The side effect is that all changes are stored in the JIP queue until the arsenal or its object is deleted. I'm not sure how to handle this any better.

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/arsenal/functions/fnc_openBox.sqf [hunk 1]
```diff
@@ -46,6 +53,7 @@ if (isNull _displayToUse || {!isNil QGVAR(camera)}) exitWith {
     [LLSTRING(CantOpenDisplay), false, 5, 1] call EFUNC(common,displayText);
 };
 
+GVAR(center) = _center;
```
Comment (LinkIsGrim) [inline addons/arsenal/functions/fnc_openBox.sqf:56]: Why?

**Review by LinkIsGrim — COMMENTED** [review]
This should work (I can't test for a while), but I'm not a fan.

Did this work properly in 3.15.2?

File: addons/arsenal/functions/fnc_openBox.sqf (hunk 1 above)
Comment (johnb432) [reply addons/arsenal/functions/fnc_openBox.sqf:56]: Drive-by change, I just moved the definition up by several lines to group the definitions together.

**Comment by johnb432** [comment]
> This should work (I can't test for a while), but I'm not a fan.

I'm not a huge fan because of the possible growing JIP queue, but this methods technically allows the most flexibility. What are your concerns?

> Did this work properly in 3.15.2?

Given the setup wasn't that much different from what is now in 3.16.1, I think this has always been broken. But at the same time, I would be surprised if that were really the case, someone would have certainly complained about it sooner.

https://github.com/acemod/ACE3/pull/9040/files#diff-be69c81d7a52a673c5725994e6c9d11c944400be668ce2d33b9290e2e84ecbe2 #9040 did some minor changes to `FUNC(initBox)`, but I don't think those are to blame for it being broken.

I'm going to check if it was working in 3.15.2 later today.

**Comment by LinkIsGrim** [comment]
> I'm not a huge fan because of the possible growing JIP queue, but this methods technically allows the most flexibility. What are your concerns?

JIP queue is the only concern. Ideally there wouldn't be enough changes made over the course of a synced arsenal's lifetime for it to be an issue, but it *can* have an impact when a player joins. Whether it's significant enough to worry about, I don't know. I'm not worried about desync, though.

**Comment by LinkIsGrim** [comment]
If it didn't work in 3.15.2 then this is good enough of a fix. Won't approve because I can't test, but code LGTM.

**Comment by johnb432** [comment]
> I'm not sure how to go about it in a better way unfortunately.

Better way was found...

File: addons/arsenal/functions/fnc_openBox.sqf [hunk 2]
```diff
@@ -18,6 +18,11 @@
  * Public: Yes
 */
 
+// Only run this after FUNC(initBox) has run (easier to check if the settings are initialized)
+if !(EGVAR(common,settingsInitFinished)) exitWith {
+    EGVAR(common,runAtSettingsInitialized) pushBack [FUNC(openBox), _this];
+};
```
Comment (LinkIsGrim) [inline addons/arsenal/functions/fnc_openBox.sqf:24]: Is this change still necessary?

File: addons/arsenal/functions/fnc_attributeInit.sqf [hunk 3]
```diff
@@ -24,14 +24,12 @@ if (_mode > 0) then {
...
-        [{
-            params ["_object", "_items"];
-
-            [_object, _items, true] call FUNC(removeVirtualItems);
-        }, _this] call CBA_fnc_execNextFrame;
-    }, [_object, _items]] call CBA_fnc_execNextFrame;
+        !isNil {(_this select 0) getVariable QGVAR(virtualItems)}
+    }, {
+        [_this select 0, _this select 1, true] call FUNC(removeVirtualItems);
+    }, [_object, _items], 20] call CBA_fnc_waitUntilAndExecute; // 20s timeout in case of failure
```
Comment (LinkIsGrim) [inline addons/arsenal/functions/fnc_attributeInit.sqf:32]: 🔥

Comment (johnb432) [reply addons/arsenal/functions/fnc_attributeInit.sqf:32]: Not sure if you are saying it's lit or if it should be set on fire...

File: addons/arsenal/functions/fnc_openBox.sqf (hunk 2 above)
Comment (johnb432) [reply addons/arsenal/functions/fnc_openBox.sqf:24]: I'll need to take a good look at it again.

File: addons/arsenal/functions/fnc_attributeInit.sqf (hunk 3 above)
Comment (LinkIsGrim) [reply addons/arsenal/functions/fnc_attributeInit.sqf:32]: the-road-to-el-dorado-both

File: addons/arsenal/functions/fnc_openBox.sqf (hunk 2 above)
Comment (johnb432) [reply addons/arsenal/functions/fnc_openBox.sqf:24]: My testing suggests it isn't needed anymore.

**Comment by johnb432** [comment]
The arsenal now refreshes globally if `FUNC(addVirtualItems)` or `FUNC(removeVirtualItems)` was applied globally.

**Review by LinkIsGrim — APPROVED** [review]
